llvm/lib/Target/X86/X86FrameLowering.cpp

```diff
@@ -2194,13 +2194,35 @@ void X86FrameLowering::emitPrologue(MachineFunction &MF,
   // flag (DF in EFLAGS register). Clear this flag by creating "cld" instruction
   // in each prologue of interrupt handler function.
   //
-  // FIXME: Create "cld" instruction only in these cases:
+  // Create "cld" instruction only in these cases:
   // 1. The interrupt handling function uses any of the "rep" instructions.
   // 2. Interrupt handling function calls another function.
   //
-  if (Fn.getCallingConv() == CallingConv::X86_INTR)
-    BuildMI(MBB, MBBI, DL, TII.get(X86::CLD))
-        .setMIFlag(MachineInstr::FrameSetup);
+  if (Fn.getCallingConv() == CallingConv::X86_INTR) {
+    bool NeedsCLD = false;
+
+    // Check if the function calls another function.
+    for (const BasicBlock &BB : Fn) {
+      for (const Instruction &I : BB) {
+        if (isa<CallInst>(I)) {
+          NeedsCLD = true;
+          break;
+        }
+
+        // Check for rep opcode.
+        unsigned Opcode = I.getOpcode();
+        if (X86::REP_MOVSB_32 <= Opcode && Opcode <= X86::REP_STOSW_64) {
```

Comment:
I'm not sure I like the range check here. At the very least can you static_assert the expected size of the range in case something is added in the future that shouldn't be here.

Reply:
Maybe add a isREPInstruction(MachineInstr &MI) in X86InstrInfo.h/cpp ?

Reply:
If someone adds an instruction in the future like `REP_NEW`, it will end up in the middle of this range. I want this code to be audited if that happens. So please add a static_assert that we currently expect 16 instructions between REP_MOVSB_32 and REP_STOSW_64. That way the static_assert will fail if a new instruction is added the breaks the range. The person making the change will see the error and update the code if the new instruction should not or should not be included.